fuel-bft/src/message.rs

```diff
@@ -1,25 +1,73 @@
-use crate::{Block, HeightRound, Key, State, ValidatorId};
+use crate::{HeightRound, RoundValidator, State};
 
-pub trait Message {
-    type Block: Block;
-    type Key: Key;
-    type ValidatorId: ValidatorId;
+use fuel_crypto::{Keystore, Message as CryptoMessage, PublicKey, Signature, Signer};
+use fuel_types::Bytes32;
 
-    fn new(round: HeightRound, key: Self::Key, state: State, block: Self::Block) -> Self;
+/// Approval of a node for a given step of the consensus round.
+pub trait Message: Send + Sync + Sized {
+    /// Block implementation of the message.
+    type Block: Default + Clone + Send + Sync;
 
+    /// Concrete error implementation for malformed requests.
+    type Error: From<<Self::Signer as Signer>::Error>;
+
+    /// Signature and keys provider.
+    type Signer: Signer;
+
+    /// Author of the message for a given round, identified by a [`PublicKey`].
+    fn author(&self) -> &RoundValidator;
+
+    /// Payload of the message.
     fn block(&self) -> &Self::Block;
-    fn owned_block(&self) -> Self::Block;
-    fn round(&self) -> &HeightRound;
+
+    /// Hash representation of the message. Used as signature argument.
+    fn hash(&self) -> Bytes32;
+
+    /// Accept the signature arguments, replacing the internal attributes.
+    fn set_signature(&mut self, author: PublicKey, signature: Signature);
+
+    /// Return the signature of the message.
+    fn signature(&self) -> &Signature;
```

Comment:
Would be better if this was an option, rather than relying on default value to tell whether it is unsigned or not.

Ideally we would have separate types to represent the unsigned components of a message vs a signed message.

Reply:
Message signature is mandatory for anything sent via network. I could alternatively create a different structure `UnsignedMessage`. But anyway, the message structure is being refactored as of the moment.